packages/app/src/components/PatternTable.tsx

```diff
@@ -0,0 +1,258 @@
+      const micropip = pyodide.pyimport('micropip');
+      const url = new URL(
+        '/drain3-0.9.11-py3-none-any.whl',
+        window.location.origin,
+      );
+      await micropip.install(url.href);
+      return pyodide;
+    },
+    refetchOnWindowFocus: false,
+    refetchOnMount: false,
+    refetchInterval: false,
+  });
+}
+
+async function mineEventPatterns(logs: string[], pyodide: any) {
+  pyodide.globals.set('HDXLOGS', logs);
+  return JSON.parse(
+    await pyodide.runPythonAsync(`
+import js
+import json
+from drain3 import TemplateMiner
+from drain3.template_miner_config import TemplateMinerConfig
+m = TemplateMiner(None, TemplateMinerConfig())
+results = []
+for log in HDXLOGS:
+    results.append(m.add_log_message(log))
+
+json.dumps(results)
+    `),
+  );
+}
+
+const PATTERN_COLUMN_ALIAS = '__hdx_pattern_field';
+const TIMESTAMP_COLUMN_ALIAS = '__hdx_timestamp';
+
+const useMinePatterns = ({
+  config,
+  samples,
+  bodyValueExpression,
+}: {
+  config: ChartConfigWithDateRange;
+  samples: number;
+  bodyValueExpression: string;
+}) => {
+  const configWithPrimaryAndPartitionKey = useConfigWithPrimaryAndPartitionKey({
+    ...config,
+    // TODO: User-configurable pattern columns and non-pattern/group by columns
+    select: [
+      `${bodyValueExpression} as ${PATTERN_COLUMN_ALIAS}`,
+      `${config.timestampValueExpression} as ${TIMESTAMP_COLUMN_ALIAS}`,
+    ].join(','),
+    // TODO: Proper sampling
+    orderBy: [{ ordering: 'DESC', valueExpression: 'rand()' }],
+    limit: { limit: samples },
+  });
+
+  const { data: sampleRows } = useQueriedChartConfig(
+    configWithPrimaryAndPartitionKey ?? config, // `config` satisfying type, never used due to `enabled` check
+    { enabled: configWithPrimaryAndPartitionKey != null },
+  );
+
+  const { data: pyodide } = usePyodide();
+
+  return useQuery({
+    queryKey: ['patterns', config],
+    queryFn: () => {
+      if (configWithPrimaryAndPartitionKey == null) {
+        throw new Error('Unexpected configWithPrimaryAndPartitionKey is null');
+      }
+
+      const logs =
+        sampleRows?.data.map(row => {
+          return stripAnsi(row[PATTERN_COLUMN_ALIAS] as string);
+        }) ?? [];
+
+      // patternId -> count, {bucket:count}, pattern
+      return mineEventPatterns(logs, pyodide).then(result => {
+        const rowsWithPatternId = [];
+        for (let i = 0; i < result.length; i++) {
+          const r = result[i];
+          const row = sampleRows?.data[i];
+          rowsWithPatternId.push({
+            ...row,
+            __hdx_patternId: r.cluster_id,
+            __hdx_pattern: r.template_mined,
+          });
+        }
+
+        return {
+          ...sampleRows,
+          data: rowsWithPatternId,
+          additionalKeysLength:
+            configWithPrimaryAndPartitionKey.additionalKeysLength,
+        };
+      });
+    },
+    refetchOnWindowFocus: false,
+    enabled: sampleRows != null && pyodide != null,
+  });
```

Comment:
Just noting: If JSON Parse fails in `minEventPatterns` we aren't catching the error. We probably don't _really_ care at this point, but we may want to do something about it.

Reply:
I believe react query will actually catch the error and we can in theory add a global error handler or use the specific error property returned from react query to handle it here too